libwav/files/sndwav.c

```diff
@@ -0,0 +1,407 @@
+
+typedef struct {
+    /* The buffer on the AICA side */
+    snd_stream_hnd_t shnd;
+
+    /* We either read the wav data from a file or 
+       we read from a buffer */
+    file_t wave_file;
+    const uint8_t *wave_buf;
+
+    /* Contains the buffer that we are going to send
+       to the AICA in the callback.  Should be 32-byte
+       aligned */
+    uint8_t *drv_buf;
+
+    /* Status of the stream that can be started, stopped
+       paused, ready. etc */
+    volatile int status;
+
+    snddrv_cb callback;
+
+    uint32_t loop;
+    uint32_t vol;         /* 0-255 */
+
+    uint32_t format;      /* Wave format */
+    uint32_t channels;    /* 1-Mono/2-Stereo */
+    uint32_t sample_rate; /* 44100Hz */
+    uint32_t sample_size; /* 4/8/16-Bit */
+
+    /* Offset into the file or buffer where the audio 
+       data starts */
+    uint32_t data_offset;
+
+    /* The length of the audio data */
+    uint32_t data_length;
+
+    /* Used only in reading wav data from a buffer 
+       and not a file */
+    uint32_t buf_offset;  
+   
+} snddrv_hnd;
+
+static snddrv_hnd streams[SND_STREAM_MAX];
+static volatile int sndwav_status = SNDDRV_STATUS_NULL;
+
+static void *sndwav_thread(void);
+static void *wav_file_callback(snd_stream_hnd_t hnd, int req, int *done);
+static void *wav_buf_callback(snd_stream_hnd_t hnd, int req, int *done);
+
+int wav_init(void) {
+    int i;
+
+    if(snd_stream_init() < 0)
+        return 0;
+
+    for(i = 0; i < SND_STREAM_MAX; i++) {
+        streams[i].shnd = SND_STREAM_INVALID;
+        streams[i].vol = 240;
+        streams[i].status = SNDDEC_STATUS_NULL;
+        streams[i].callback = NULL;
+    }
+
+    if(thd_create(1, sndwav_thread, NULL) != NULL) {
+        sndwav_status = SNDDRV_STATUS_READY;
+        return 1;
+	}
+    else {
+        return 0;
+    }
+}
+
+void wav_shutdown(void) {
+    int i;
+
+    sndwav_status = SNDDRV_STATUS_DONE;
+    for(i = 0; i < SND_STREAM_MAX; i++) {
+        wav_destroy(i);
+    }
+}
+
+void wav_destroy(wav_stream_hnd_t hnd) {
+    if(streams[hnd].shnd == SND_STREAM_INVALID)
+        return;
+
+    if(streams[hnd].wave_file != FILEHND_INVALID)
+        fs_close(streams[hnd].wave_file);
+
+    if(streams[hnd].drv_buf) {
+        free(streams[hnd].drv_buf);
+        streams[hnd].drv_buf = NULL;
+    }
+
+    snd_stream_stop(streams[hnd].shnd);
+    snd_stream_destroy(streams[hnd].shnd);
+    
+    streams[hnd].shnd = SND_STREAM_INVALID;
+    streams[hnd].vol = 240;
+    streams[hnd].status = SNDDEC_STATUS_NULL;
+    streams[hnd].callback = NULL;
+}
+
+wav_stream_hnd_t wav_create(const char *filename, int loop) {
+    int fn_len;
+    file_t file;
+    WavFileInfo info;
+    wav_stream_hnd_t index;
+
+    if(filename == NULL)
+        return SND_STREAM_INVALID;
+
+    file = fs_open(filename, O_RDONLY);
+
+    if(file == FILEHND_INVALID)
+        return SND_STREAM_INVALID;
+
+    index = snd_stream_alloc(wav_file_callback, SND_STREAM_BUFFER_MAX); // SND_STREAM_BUFFER_MAX/4
```

Comment:
This comment a bit confused:
"// SND_STREAM_BUFFER_MAX/4"

Reply:
It's a stub holder. I guess I can remove it.